src/pyggdrasil/scripts/cell_simulation.py

```diff
@@ -0,0 +1,369 @@
+    adj_matrix = np.zeros((n_nodes, n_nodes))
+    # Generate random edges for the tree
+    for i in range(1, n_nodes):
+        # Select a random parent node from previously added nodes
+        parent = random.choice(rng, i)
+        # Add an edge from the parent to the current node
+        adj_matrix[parent, i] = 1
+    # Return the adjacency matrix
+    return adj_matrix
+
+
+def reverse_node_order(adj_matrix: np.ndarray) -> np.ndarray:
+    """
+    Reverses the order of the nodes in the tree adjacency matrix.
+    Args:
+        adj_matrix: np.ndarray
+            adjacency matrix
+
+    Returns:
+        adj_matrix: np.ndarray
+            adjacency matrix
+    """
+    # Reverse the order of the nodes
+    adj_matrix = adj_matrix[::-1, ::-1]
+    # Return the adjacency matrix
+    return adj_matrix
+
+
+def compose_save_name(params: dict, *, tree_no: int) -> str:
+    """Composes save name for the results."""
+    save_name = (
+        f"seed_{params['seed']}_"
+        f"n_trees_{params['n_trees']}_"
+        f"n_cells_{params['n_cells']}_"
+        f"n_mutations_{params['n_mutations']}_"
+        f"alpha_{params['alpha']}_"
+        f"beta_{params['beta']}_"
+        f"na_rate_{params['na_rate']}_"
+        f"observe_homozygous_{params['observe_homozygous']}_"
+        f"strategy_{params['strategy']}"
+    )
+    if tree_no is not None:
+        save_name += f"_tree_{tree_no}"
+    return save_name
+
+
+def adjacency_to_root_dfs(adj_matrix: np.ndarray) -> TreeNode:
+    """Convert adjacency matrix to tree in tree.TreeNode
+        traverses a tree using depth first search.
+
+    Args:
+        adj_matrix: np.ndarray
+            with no self-loops (i.e. diagonal is all zeros)
+            and the root as the highest index node
+    Returns:
+        root: TreeNode containing the entire tree
+    """
+
+    # Determine the root node (node with highest index)
+    root_idx = len(adj_matrix) - 1
+
+    # Create a stack to keep track of nodes to visit
+    stack = [root_idx]
+
+    # Create a set to keep track of visited nodes
+    visited = set()
+
+    # Create a list to keep track of nodes
+    child_parent = {}
+
+    # Create a list to keep track of TreeNodes
+    list_TreeNode = np.empty(len(adj_matrix), dtype=TreeNode)
+
+    # Traverse the tree using DFS
+    while stack:
+        # Get the next node to visit
+        node = stack.pop()
+
+        # Skip if already visited
+        if node in visited:
+            # print(f"Already Visited node {node}")
+            continue
+
+        # Visit the node
+        # print(f"Visiting node {node}")
+
+        # print(f"Parent of node {node} is {child_parent[node]}")
+
+        if node == root_idx:
+            root = TreeNode(name=node, data=None, parent=None)
+            list_TreeNode[node] = root
+        else:
+            # Recall parent
+            parent = child_parent[node]
+            child = TreeNode(name=node, data=None, parent=list_TreeNode[parent])
```

Comment:
pyright has been giving me a hard time here. This line gives a type error, which I have so far failed to resolve. See the error message below. Any idea?

`list_TreeNode` is an initially empty list of TreeNode elements that is filled with the actual nodes of the tree by transversing the tree using depth-search search. `parent` is just the `int` index of a node in that list.

> cell_simulation.py:194:59 - error: Argument of type "ndarray[Any, dtype[Unknown]]" cannot be assigned to parameter "parent" of type "TreeNode[Unknown, Unknown] | None" in function "__init__"
    Type "ndarray[Any, dtype[Unknown]]" cannot be assigned to type "TreeNode[Unknown, Unknown] | None"
      "ndarray[Any, dtype[Unknown]]" is incompatible with "TreeNode[Unknown, Unknown]"
      Type cannot be assigned to type "None"

Reply:
Ah, I see! `pyright` has a problem with inferring the type of elements stored in the NumPy array, so this is just false positive.

Adding `  # type: ignore` comment to line 194 should silence the error.